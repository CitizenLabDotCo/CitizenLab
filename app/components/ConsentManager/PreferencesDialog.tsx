import React, { PureComponent } from 'react';

// Components
import CategoryCard from './CategoryCard';

// Typing
import { IDestination } from './';

// Styling
import styled from 'styled-components';
import { fontSizes, media } from 'utils/styleUtils';

export const ContentContainer = styled.div`
  padding: 30px;
  background: white;

  h1 {
    font-size: ${fontSizes.medium}px;
  }

  h2 {
    font-size: ${fontSizes.large}px;
  }

  ${media.smallerThanMaxTablet`
    margin: 0;
    padding: 20px;
  `}
`;

interface Props {
  onChange: (category, value) => void;
  categoryDestinations: {
    analytics: IDestination[];
    advertising: IDestination[];
    functional: IDestination[];
  };
  analytics: boolean | null;
  advertising: boolean | null;
  functional: boolean | null;
}

export default class PreferencesDialog extends PureComponent<Props> {
  static displayName = 'PreferencesDialog';

  handleChange = (event) => {
    event.preventDefault();
    this.props.onChange(event.target.name, event.target.value === 'true');
  }

  render() {
    const {
      categoryDestinations,
      functional,
      advertising,
      analytics,
    } = this.props;
    const checkCategories = { analytics, advertising, functional };
    return (
      <ContentContainer id="e2e-preference-dialog">
        {Object.keys(categoryDestinations).map((category) => {
          if (categoryDestinations[category].length > 0) {
            return (
              <CategoryCard
                key={category}
                category={category}
                destinations={categoryDestinations[category]}
                checked={checkCategories[category]}
                handleChange={this.handleChange}
              />
            );
          }
          return;
        })}
      </ContentContainer>
    );
  }
}
