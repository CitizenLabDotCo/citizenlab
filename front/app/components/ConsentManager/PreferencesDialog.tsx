import React from 'react';

// Components
import CategoryCard from './CategoryCard';

// Typing
import { CategorizedDestinations, IPreferences } from '.';

// Styling
import styled from 'styled-components';
import { fontSizes, media } from 'utils/styleUtils';

// services
import { TCategory } from './destinations';

export const ContentContainer = styled.div`
  padding: 30px;
  background: white;

  h1 {
    font-size: ${fontSizes.m}px;
  }

  h2 {
    font-size: ${fontSizes.l}px;
  }

  ${media.phone`
    margin: 0;
    padding: 20px;
  `}
`;

interface Props {
  onChange: (category: TCategory, value) => void;
  categoryDestinations: CategorizedDestinations;
  preferences: IPreferences;
}

const doNothing = () => () => {};

const PreferencesDialog = ({
  categoryDestinations,
  preferences,
  onChange,
}: Props) => {
  const handleChange = (category: TCategory, value: boolean) => (_event) => {
    onChange(category, value);
  };

  return (
    <ContentContainer id="e2e-preference-dialog">
      {Object.keys(categoryDestinations)
        .filter((category: TCategory) => {
          return categoryDestinations[category].length > 0;
        })
        .map((category: TCategory) => {
          return (
            <CategoryCard
              key={category}
              category={category}
              destinations={categoryDestinations[category]}
              checked={!!preferences[category]}
              handleChange={handleChange}
            />
          );
        })}
      <CategoryCard
        key={'required'}
        category={'required'}
        destinations={[]}
        checked={true}
        handleChange={doNothing}
        disableUncheck={true}
      />
    </ContentContainer>
  );
};

export default PreferencesDialog;
