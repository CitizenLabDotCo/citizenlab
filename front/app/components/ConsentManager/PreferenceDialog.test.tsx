// @ts-nocheck
// libraries
import React from 'react';
import { shallow } from 'enzyme';

import 'jest-styled-components';

// component to test
import PreferencesDialog from './PreferencesDialog';

// mock utilities
jest.mock('utils/cl-intl');
jest.mock('services/appConfiguration');
jest.mock('modules', () => ({ streamsToReset: [] }));

describe('<ConsentManager />', () => {
  let onChange: Jest.Mock;

  const categoryDestinations = {
    analytics: [
      {
        name: 'Google Analytics',
        description:
          'Google Analytics is the most popular marketing tool for the web. It’s free and provides a wide range of features. It’s especially good at measuring traffic sources and ad campaigns.',
        category: 'Analytics',
        website: 'http://google.com/analytics',
        id: 'Google Analytics',
      },
      {
        name: 'MarketingTool',
        description:
          'MarketingTool is the most popular marketing tool for the web. It’s free and provides a wide range of features. It’s especially good at measuring traffic sources and ad campaigns.',
        category: 'Tag Managers',
        website: 'http://random.com/marketing',
        id: 'MarketingTool',
      },
    ],
    advertising: [
      {
        name: 'AdvertisingTool',
        description: 'Advertising BS',
        category: 'Advertising',
        website: 'http://random.com/advertising',
        id: 'AdvertisingTool',
      },
    ],
    functional: [
      {
        name: 'FunctionalTool',
        description: 'Actually might be handy',
        category: 'Security & Fraud',
        website: 'http://random.com/securitycookie',
        id: 'FunctionalTool',
      },
    ],
  };
  const preferences = {
    analytics: true,
    advertising: null,
    functional: null,
  };

  beforeEach(() => {
    onChange = jest.fn();
  });

  it('renders correctly when there are destinations', () => {
    const wrapper = shallow(
      <PreferencesDialog
        onChange={onChange}
        categoryDestinations={categoryDestinations}
        preferences={preferences}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
  it('renders correctly when there are no destinations', () => {
    const wrapper = shallow(
      <PreferencesDialog
        onChange={onChange}
        categoryDestinations={{
          analytics: [],
          advertising: [],
          functional: [],
        }}
        preferences={preferences}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
});
