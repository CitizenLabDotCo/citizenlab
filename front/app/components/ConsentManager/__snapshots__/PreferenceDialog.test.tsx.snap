// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ConsentManager /> renders correctly when there are destinations 1`] = `
<PreferencesDialog__ContentContainer
  id="e2e-preference-dialog"
>
  <CategoryCard
    category="analytics"
    checked={true}
    destinations={
      [
        {
          "category": "Analytics",
          "description": "Google Analytics is the most popular marketing tool for the web. It’s free and provides a wide range of features. It’s especially good at measuring traffic sources and ad campaigns.",
          "id": "Google Analytics",
          "name": "Google Analytics",
          "website": "http://google.com/analytics",
        },
        {
          "category": "Tag Managers",
          "description": "MarketingTool is the most popular marketing tool for the web. It’s free and provides a wide range of features. It’s especially good at measuring traffic sources and ad campaigns.",
          "id": "MarketingTool",
          "name": "MarketingTool",
          "website": "http://random.com/marketing",
        },
      ]
    }
    handleChange={[Function]}
    key="analytics"
  />
  <CategoryCard
    category="advertising"
    checked={false}
    destinations={
      [
        {
          "category": "Advertising",
          "description": "Advertising BS",
          "id": "AdvertisingTool",
          "name": "AdvertisingTool",
          "website": "http://random.com/advertising",
        },
      ]
    }
    handleChange={[Function]}
    key="advertising"
  />
  <CategoryCard
    category="functional"
    checked={true}
    destinations={
      [
        {
          "category": "Security & Fraud",
          "description": "Actually might be handy",
          "id": "FunctionalTool",
          "name": "FunctionalTool",
          "website": "http://random.com/securitycookie",
        },
      ]
    }
    handleChange={[Function]}
    key="functional"
  />
  <CategoryCard
    category="required"
    checked={true}
    destinations={[]}
    disableUncheck={true}
    handleChange={[Function]}
    key="required"
  />
</PreferencesDialog__ContentContainer>
`;

exports[`<ConsentManager /> renders correctly when there are no destinations 1`] = `
<PreferencesDialog__ContentContainer
  id="e2e-preference-dialog"
>
  <CategoryCard
    category="required"
    checked={true}
    destinations={[]}
    disableUncheck={true}
    handleChange={[Function]}
    key="required"
  />
</PreferencesDialog__ContentContainer>
`;
